### PR TITLE
provider/ec2: update default controlelr instance

### PR DESCRIPTION
**Pull request by axw**
In a VPC account, we'll create a t2.medium for
controller instances by default now. Non-controller
instances are unchanged for now, and remain m3.medium
by default. Controller instance root disk now defaults
to 32GiB, but can be overridden with constraints.

There is a fix to the ec2instancetypes generator:
we were expecting "currentGeneration" to be present
in the attributes always, but it is not there for
some. We assume they are current if it is missing,
which is the case with the ones that have it missing
currently.

Fixes https://bugs.launchpad.net/juju/+bug/1535838

Requires https://github.com/go-amz/amz/pull/71

**QA**

bootstrap us-east-1, with no default VPC:
 - m3.medium with 32G root disk
add-machine:
 - m3.medium with 8G root disk

bootstrap ap-southeast-2, with default VPC:
 - t2.medium with 32G root disk
add-machine:
 - m3.medium with 8G root disk
add-machine --constraints cpu-power=40
 - t2.medium with 8G root disk

bootstrap ap-southeast-2, with default VPC,
and --constraints root-disk=8G:
 - t2.medium with 8G root disk

### CODE REVIEW

File: constraints/constraints.go [hunk 1]
```diff
@@ -110,6 +110,18 @@ func (v *Value) HasArch() bool {
 	return v.Arch != nil && *v.Arch != ""
 }
 
+func (v *Value) HasMem() bool {
```
Comment (wallyworld) [inline constraints/constraints.go]: These need comments

File: provider/ec2/image.go [hunk 2]
```diff
@@ -44,20 +37,53 @@ func filterImages(images []*imagemetadata.ImageMetadata, ic *instances.InstanceC
...
 }
+
+// withDefaultControllerConstraints returns the given constraints,
+// updated to choose a default instance type appropriate for a
+// controller machine. We use this only if the user does not specify
+// any constraints that would otherwise control the instance type
+// selection.
+//
+// At the time of writing, this will choose
+//   - t2.medium, for VPC
```
Comment (wallyworld) [inline provider/ec2/image.go]: This is probably all fine. Just to check - Uros has a concern about controllers being CPU bound. Is t2.medium sufficient? Is it correct not to require an instance with a specified minimum cpu-power figure?

File: constraints/constraints.go (hunk 1 above)
Comment (axw) [inline constraints/constraints.go]: Done.

File: provider/ec2/image.go (hunk 2 above)
Comment (axw) [inline provider/ec2/image.go]: This is just a default. If you're expecting to manage a large environment, then you should specify a higher minimum.

We did talk about specifying an instance type by name, but the problem with that is that when T3 inevitably comes along, we won't select it without further code changes. We don't yet auto-update instance types, but my recent changes should make it fairly easy to do so.

**Comment by axw** [comment]
$$merge$$

**Pull request merged**